File: leetcode/python/19. Remove Nth Node From End of List.py
class ListNode:
  
  def __init__(self,val=0):
    self.val=val
    self.next=None
  
class Main:

  def __init__(self):
    self.root = None

  def add(self,data):
    node = ListNode(data)
    if self.root is None:
      self.root = node
    else:
      temp = self.root
      while temp.next is not None:
        temp = temp.next
      temp.next = node
  
  def removeNthFromEnd(self,n):
    count = 0
    tmp = self.root
    while tmp is not None:
      tmp = tmp.next
      count+=1
    toDe = count - n
    parent = ListNode(0)
    parent.next = self.root
    second = parent
    while toDe > 0:
      second = second.next
      toDe -= 1
    second.next = second.next.next
    return parent.next
  
  def display(self):
    tmp = self.root
    while tmp is not None:
      print(tmp.val)
      tmp = tmp.next

arr = list(range(6))
m = Main()
for i in range(1,len(arr)):
  m.add(arr[i])
m.display()
y = m.removeNthFromEnd(2)
while y is not None:
  print(y.val)
  y = y.next

File: leetcode/js/19. Remove Nth Node From End of List.js
class ListNode {
    constructor(val=0){ 
      this.val = val
      this.next = null
    }
  }
  
  class Main {
    
    constructor(){
      this.root=null
    }

    // static get root(){
    //   delete Main.root
    //   return Main.root
    // }

    // static set root(val){
    //   delete Main.root
    //   Main.root = val
    // }

    // get root(){
    //   delete this.root
    //   return this.root
    // }

    // set root(val){
    //   delete this.root
    //   this.root = val
    // }

    add(data){
         let node = new ListNode(data)
         if(this.root == null){
           this.root = node
         }else{
           let temp= this.root
           while(temp.next){
             temp=temp.next
           }
           temp.next=node
         }
       }
    
  
    removeNthFromEnd(n){
      var len = 0
      var tmp = this.root
      while (tmp != null) {
        tmp = tmp.next
        len++
      }
      var toDe = len - n
      var parent = new ListNode(0)
      parent.next = this.root
      var second = parent
      while (toDe > 0) {
        second = second.next
        toDe--
      }
      second.next = second.next.next
      return parent.next
    }
    
    display(temp=this.root){
      while(temp!=null){
        console.log(temp.val)
        temp=temp.next
      }
    }
  }
  
  const arr = [...Array(5).keys()].map(x=>x+1)
  
  var m = new Main()
   for (const i in arr) {
    m.add(arr[i])  
  }
  
  m.display()
  var y = m.removeNthFromEnd(2)
  
  while(y!=null){
    console.log(y.val)
    y=y.next
  }
  
  
  